Fixed "KeyError: 'last_updated'"

avityk opened the pull request:
# Description:
Recently, when syncing lists, the following error began to appear:
```
2018-11-13 07:26:21,530Z    tap -   INFO GET https://go.urbanairship.com/api/lists
2018-11-13 07:26:21,646Z    tap - CRITICAL 'last_updated'
2018-11-13 07:26:21,649Z    tap - Traceback (most recent call last):
2018-11-13 07:26:21,650Z    tap -   File "tap-env/bin/tap-urban-airship", line 11, in <module>
2018-11-13 07:26:21,650Z    tap -     sys.exit(main())
2018-11-13 07:26:21,651Z    tap -   File "/code/orchestrator/tap-env/lib/python3.5/site-packages/tap_urban_airship/__init__.py", line 136, in main
2018-11-13 07:26:21,651Z    tap -     raise exc
2018-11-13 07:26:21,651Z    tap -   File "/code/orchestrator/tap-env/lib/python3.5/site-packages/tap_urban_airship/__init__.py", line 133, in main
2018-11-13 07:26:21,651Z    tap -     main_impl()
2018-11-13 07:26:21,652Z    tap -   File "/code/orchestrator/tap-env/lib/python3.5/site-packages/tap_urban_airship/__init__.py", line 129, in main_impl
2018-11-13 07:26:21,652Z    tap -     do_sync()
2018-11-13 07:26:21,652Z    tap -   File "/code/orchestrator/tap-env/lib/python3.5/site-packages/tap_urban_airship/__init__.py", line 105, in do_sync
2018-11-13 07:26:21,653Z    tap -     sync_entity("lists", ["name"], ["created", "last_updated"])
2018-11-13 07:26:21,653Z    tap -   File "/code/orchestrator/tap-env/lib/python3.5/site-packages/tap_urban_airship/__init__.py", line 89, in sync_entity
2018-11-13 07:26:21,653Z    tap -     last_touched = max(row[date_key] for date_key in date_keys)
2018-11-13 07:26:21,654Z    tap -   File "/code/orchestrator/tap-env/lib/python3.5/site-packages/tap_urban_airship/__init__.py", line 89, in <genexpr>
2018-11-13 07:26:21,654Z    tap -     last_touched = max(row[date_key] for date_key in date_keys)
2018-11-13 07:26:21,654Z    tap - KeyError: 'last_updated'
```
# The cause of the problem:
The row does not always have the `last_updated` field, but only the `created` field.

# Solution:
Added a check for key presence.
Now it runs normally:
```
  INFO Starting sync
  INFO GET https://go.urbanairship.com/api/lists
{"stream": "lists", "schema": {"type": "object", "properties": {"name": {"type": "string"}, "description": {"type": ["string", "null"]}, "extra": {"anyOf": [{"type": "null"}, {"type": "array", "items": {"type": "object", "properties": {"key": {"type": "string"}, "value": {"type": ["string", "null"]}}}}]}, "created": {"type": ["string", "null"], "format": "date-time"}, "last_modified": {"type": ["string", "null"], "format": "date-time"}, "channel_count": {"type": ["integer", "null"]}, "status": {"type": ["string", "null"]}}}, "key_properties": ["name"], "type": "SCHEMA"}
{"stream": "lists", "record": {"name": "Test List", "description": "Test Users", "extra": [], "created": "2018-11-12T17:05:37Z", "channel_count": 2, "status": "ready"}, "type": "RECORD"}
{"value": {"lists": "2018-11-12T17:05:37Z"}, "type": "STATE"}
```

avityk committed:
Fixed "KeyError: 'last_updated'"

timvisher commented:
It's a little weird to me that the 'cause of the problem' here doesn't seem to relate directly to the code change. The `last_updated`, `created` entity seems to be `lists` whereas the fix appears to be fixing the misspelling of `modification_date`?

Could you maybe update the PR description to reflect that or explain the discrepancy?

avityk commented:
@timvisher The fix is in this row:
```
last_touched = max(row[date_key] for date_key in date_keys)
->
last_touched = max(row[date_key] for date_key in date_keys if date_key in row)
```
If `last_updated` field was not present in the row, `KeyError` was raised. So I added checking for key presence. If this key is missing, only `created` field is being checked.
If for some reason the `created` field was also absent, then an `ValueError: max() arg is an empty sequence` error would appear.  So I added these rows before checking the date:
```
if not any(date_key in row for date_key in date_keys):
    raise KeyError('None of date keys found in the row')
```
So in this case error message will be more informative.

And fixing the misspelling of `modification_date` is not related to this issue, I just noticed it and corrected.

luandy64 commented on tap_urban_airship/__init__.py:
```diff
@@ -104,7 +106,7 @@ def do_sync():
     sync_entity("lists", ["name"], ["created", "last_updated"])
     sync_entity("channels", ["channel_id"], ["created", "last_registration"])
-    sync_entity("segments", ["id"], ["creation_date", "modificiation_date"])
+    sync_entity("segments", ["id"], ["creation_date", "modification_date"])
```
Given that this wasn't part of the fix the PR was intended to make, let's remove this change and potentially open up a separate PR to address it.

On that PR, I'm somewhat inclined to delete it rather than fix the spelling of it given that it can't possibly have ever been used before in its current state.

avityk replied:
I rolled back this change from this PR.

luandy64 commented on tap_urban_airship/__init__.py:
```diff
@@ -86,7 +86,9 @@ def sync_entity(entity, primary_keys, date_keys=None, transform=None):
         row = transform_row(row, schema)
         if date_keys:
-            last_touched = max(row[date_key] for date_key in date_keys)
+            if not any(date_key in row for date_key in date_keys):
```
At first glance, it seems like if we are getting key errors here then we need to adjust how we call this function.

If I'm wrong here, could you add some comments to clarify your logic here?

avityk replied:
Ideally, this error will never be raised. But at the moment we are having an issue caused by missing `last_updated` key in the row. So I added a `KeyError` raising here in order to, if suddenly there is a second key missing, we noticed it instead of unclear `ValueError: max() arg is an empty sequence`.

luandy64 replied:
Ah that makes sense.

I think the various list comprehensions here are complicated enough that it would warrant a comment. Probably right above line 88.

Something along the lines of:

```
# Rows can have various values for various sets of keys (See the calls to
# `sync_entitity` in `do_sync`).
#
# To handle this we:
#
# 1. Get _all_ the values for all the keys that are actually present in
# the row (not every row has every key).
#
# 2. Take the max of those values as the bookmark for that entity.
#
# A KeyError is raised in the event that a row has none of the bookmark
# keys.
```

avityk replied:
Done.

luandy64 commented:
@avityk Just to clarify, is this an issue only on the `lists` endpoint, or do you see this for other endpoints as well?

avityk commented:
@luandy64 No, I haven't seen this issue for other endpoints.

luandy64 reviewed (changes requested):
Once we add a comment can we get an updated gist of an example run verifying that things are looking good? Then I think we should be all good to ship this. Thanks!

avityk commented:
@luandy64 Here is the log snippet:
```
  INFO Starting sync
  INFO GET https://go.urbanairship.com/api/lists
{"stream": "lists", "schema": {"type": "object", "properties": {"name": {"type": "string"}, "description": {"type": ["string", "null"]}, "extra": {"anyOf": [{"type": "null"}, {"type": "array", "items": {"type": "object", "properties": {"key": {"type": "string"}, "value": {"type": ["string", "null"]}}}}]}, "created": {"type": ["string", "null"], "format": "date-time"}, "last_modified": {"type": ["string", "null"], "format": "date-time"}, "channel_count": {"type": ["integer", "null"]}, "status": {"type": ["string", "null"]}}}, "key_properties": ["name"], "type": "SCHEMA"}
{"stream": "lists", "record": {"name": "Test List", "description": "Test Users", "extra": [], "created": "2018-11-12T17:05:37Z", "channel_count": 2, "status": "ready"}, "type": "RECORD"}
{"value": {"lists": "2018-11-12T17:05:37Z"}, "type": "STATE"}
```

luandy64 commented:
@avityk this all looks good! Can you squash this so we can get it merged?

You could then go ahead and rebase #8 onto master so we can get that going too.

avityk commented:
@luandy64 Done.

luandy64 commented:
Thanks for working on this PR! Let's move over to #8 and I'll get this released after that.
